client/core/bot.go

```diff
@@ -0,0 +1,1733 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"net/url"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/client/db"
+	"decred.org/dcrdex/client/orderbook"
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/calc"
+	"decred.org/dcrdex/dex/encrypt"
+	"decred.org/dcrdex/dex/order"
+)
+
+const (
+	// MakerBotV0 is the bot specifier associated with makerBot. The bot
+	// specifier is used to determine how to decode stored program data.
+	MakerBotV0 = "MakerV0"
+
+	oraclePriceExpiration = time.Minute * 10
+	oracleRecheckInterval = time.Minute * 3
+
+	ErrNoMarkets           = dex.ErrorKind("no markets")
+	defaultOracleWeighting = 0.2
+
+	// Our mid-gap rate derived from the local DEX order book is converted to an
+	// effective mid-gap that can only vary by up to 3% from the oracle rate.
+	// This is to prevent someone from taking advantage of a sparse market to
+	// force a bot into giving a favorable price. In reality a market maker on
+	// an empty market should use a high oracle bias anyway, but this should
+	// prevent catastrophe.
+	maxOracleMismatch = 0.03
+)
+
+// GapStrategy is a specifier for an algorithm to choose the maker bot's target
+// spread.
+type GapStrategy string
+
+const (
+	// GapStrategyMultiplier calculates the spread by multiplying the
+	// break-even gap by the specified multiplier, 1 <= r <= 100.
+	GapStrategyMultiplier GapStrategy = "multiplier"
+	// GapStrategyAbsolute sets the spread to the rate difference.
+	GapStrategyAbsolute GapStrategy = "absolute"
+	// GapStrategyAbsolutePlus sets the spread to the rate difference plus the
+	// break-even gap.
+	GapStrategyAbsolutePlus GapStrategy = "absolute-plus"
+	// GapStrategyPercent sets the spread as a ratio of the mid-gap rate.
+	// 0 <= r <= 0.1
+	GapStrategyPercent GapStrategy = "percent"
+	// GapStrategyPercentPlus sets the spread as a ratio of the mid-gap rate
+	// plus the break-even gap.
+	GapStrategyPercentPlus GapStrategy = "percent-plus"
```

Comment:
I don't have a great grasp on what is needed here, but seems like `Plus` categories could just be an option and don't need separate strategies. Also, why add the break-even? Could set to break even if gap is lower also work?

Asking for my understanding.

Reply:
Gaps above the break-even are "profitable". A gap below the break-even guarantees losses. Having break-even as a minimum allows the bot to adapt to changing market conditions, particularly changes in tx fee rates, without user input.

Reply:
> Could set to break even if gap is lower also work?

I'm not sure I understand this question.

Reply:
> I'm not sure I understand this question.

If break even is .07 and the program would set to .1 adding sets it to .17 but really .1 was already enough to bring it above.